File: android/app/src/test/java/com/recre/app/core/calculo/RecuperacionTest.kt
package com.recre.app.core.calculo

import java.math.BigDecimal
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Replica los 8 casos de `supabase/functions/_shared/recuperacion.test.ts`
 * para garantizar que el espejo Kotlin ([planificarRecuperacion]) produce el
 * mismo plan que el SSOT TypeScript que ejecuta el servidor.
 *
 * Si la lógica del servidor cambia, este test fallará y obligará a actualizar
 * el espejo cliente al mismo tiempo (igual que [CalculoTest] / [SemanasIsoTest]).
 */
class RecuperacionTest {

    private fun bd(v: String) = BigDecimal(v)

    private fun tolva(id: String, saldo: String, fecha: String = "2026-03-01") =
        CreditoAbierto(id = id, tipo = "tolva", saldo = bd(saldo), fecha = fecha)

    private fun prestamo(id: String, saldo: String, fecha: String) =
        CreditoAbierto(id = id, tipo = "prestamo", saldo = bd(saldo), fecha = fecha)

    /** Compara importes ignorando la escala (`0` ≡ `0.00`). */
    private fun assertImporte(esperado: String, real: BigDecimal) =
        assertEquals(0, bd(esperado).compareTo(real))

    @Test
    fun `pct=0 no recupera nada`() {
        val plan = planificarRecuperacion(bd("100.00"), 0, listOf(tolva("t1", "100.00")))
        assertImporte("0.00", plan.recuperadoTotal)
        assertImporte("100.00", plan.pagadoLocal)
        assertEquals(emptyList<AsignacionRecuperacion>(), plan.asignaciones)
    }

    @Test
    fun `sin deudas pagado igual a parte_local aunque pct mayor que 0`() {
        val plan = planificarRecuperacion(bd("100.00"), 100, emptyList())
        assertImporte("0.00", plan.recuperadoTotal)
        assertImporte("100.00", plan.pagadoLocal)
    }

    @Test
    fun `100 por ciento con deuda suficiente el local no se lleva nada`() {
        val plan = planificarRecuperacion(bd("100.00"), 100, listOf(tolva("t1", "100.00")))
        assertImporte("100.00", plan.recuperadoTotal)
        assertImporte("0.00", plan.pagadoLocal)
        assertEquals(1, plan.asignaciones.size)
        assertEquals("t1", plan.asignaciones[0].creditoId)
        assertImporte("100.00", plan.asignaciones[0].importe)
    }

    @Test
    fun `50 por ciento retiene la mitad`() {
        val plan = planificarRecuperacion(bd("100.00"), 50, listOf(tolva("t1", "100.00")))
        assertImporte("50.00", plan.recuperadoTotal)
        assertImporte("50.00", plan.pagadoLocal)
    }

    @Test
    fun `se topa al saldo de la deuda`() {
        val plan = planificarRecuperacion(
            bd("100.00"),
            100,
            listOf(prestamo("p1", "30.00", "2026-01-01")),
        )
        assertImporte("30.00", plan.recuperadoTotal)
        assertImporte("70.00", plan.pagadoLocal)
    }

    @Test
    fun `imputa tolva primero luego prestamo aunque el prestamo sea mas antiguo`() {
        val plan = planificarRecuperacion(
            bd("200.00"),
            100,
            listOf(
                prestamo("p1", "100.00", "2026-01-01"),
                tolva("t1", "50.00", "2026-05-01"),
            ),
        )
        assertImporte("150.00", plan.recuperadoTotal)
        assertImporte("50.00", plan.pagadoLocal)
        assertEquals(listOf("t1", "p1"), plan.asignaciones.map { it.creditoId })
        assertImporte("50.00", plan.asignaciones[0].importe)
        assertImporte("100.00", plan.asignaciones[1].importe)
    }

    @Test
    fun `FIFO entre prestamos la deuda mas antigua primero`() {
        val plan = planificarRecuperacion(
            bd("100.00"),
            100,
            listOf(
                prestamo("nuevo", "80.00", "2026-05-01"),
                prestamo("viejo", "40.00", "2026-01-01"),
            ),
        )
        assertImporte("100.00", plan.recuperadoTotal)
        assertEquals(listOf("viejo", "nuevo"), plan.asignaciones.map { it.creditoId })
        assertImporte("40.00", plan.asignaciones[0].importe)
        assertImporte("60.00", plan.asignaciones[1].importe)
    }

    @Test
    fun `orden manual antepone los creditos indicados`() {
        val plan = planificarRecuperacion(
            bd("100.00"),
            100,
            listOf(
                prestamo("nuevo", "80.00", "2026-05-01"),
                prestamo("viejo", "40.00", "2026-01-01"),
            ),
            orden = listOf("nuevo"),
        )
        assertImporte("100.00", plan.recuperadoTotal)
        assertEquals(listOf("nuevo", "viejo"), plan.asignaciones.map { it.creditoId })
        assertImporte("80.00", plan.asignaciones[0].importe)
        assertImporte("20.00", plan.asignaciones[1].importe)
    }
}

File: supabase/functions/_shared/recuperacion.test.ts
/**
 * Tests del SSOT de recuperación de deuda (T-214).
 *
 * Ejecutar con:
 *   deno test supabase/functions/_shared/recuperacion.test.ts
 *
 * Pura: sin red ni BBDD. Android replica estos mismos casos en Kotlin (T-215).
 */

import { assertEquals } from "@std/assert";

import { type CreditoAbierto, planificarRecuperacion } from "./recuperacion.ts";

const tolva = (id: string, saldo: string, fecha = "2026-03-01"): CreditoAbierto => ({
  id,
  tipo: "tolva",
  saldo,
  fecha,
});
const prestamo = (id: string, saldo: string, fecha: string): CreditoAbierto => ({
  id,
  tipo: "prestamo",
  saldo,
  fecha,
});

Deno.test("pct=0 no recupera nada: pagado = parte_local", () => {
  const plan = planificarRecuperacion({
    parteLocal: "100.00",
    porcentajeRecuperacion: 0,
    creditos: [tolva("t1", "100.00")],
  });
  assertEquals(plan.recuperado_total, "0.00");
  assertEquals(plan.pagado_local, "100.00");
  assertEquals(plan.asignaciones, []);
});

Deno.test("sin deudas: pagado = parte_local aunque pct>0", () => {
  const plan = planificarRecuperacion({
    parteLocal: "100.00",
    porcentajeRecuperacion: 100,
    creditos: [],
  });
  assertEquals(plan.recuperado_total, "0.00");
  assertEquals(plan.pagado_local, "100.00");
});

Deno.test("100% con deuda suficiente: el local no se lleva nada (ejemplo design.md §5.5)", () => {
  // bruto 260, tasa 60 → neto 200, % local 50 → parte_local 100; deuda 100.
  const plan = planificarRecuperacion({
    parteLocal: "100.00",
    porcentajeRecuperacion: 100,
    creditos: [tolva("t1", "100.00")],
  });
  assertEquals(plan.recuperado_total, "100.00");
  assertEquals(plan.pagado_local, "0.00");
  assertEquals(plan.asignaciones, [{ credito_id: "t1", importe: "100.00" }]);
});

Deno.test("50% retiene la mitad de la parte_local", () => {
  const plan = planificarRecuperacion({
    parteLocal: "100.00",
    porcentajeRecuperacion: 50,
    creditos: [tolva("t1", "100.00")],
  });
  assertEquals(plan.recuperado_total, "50.00");
  assertEquals(plan.pagado_local, "50.00");
});

Deno.test("se topa al saldo de la deuda (no retiene más de lo que se debe)", () => {
  const plan = planificarRecuperacion({
    parteLocal: "100.00",
    porcentajeRecuperacion: 100,
    creditos: [prestamo("p1", "30.00", "2026-01-01")],
  });
  assertEquals(plan.recuperado_total, "30.00");
  assertEquals(plan.pagado_local, "70.00");
});

Deno.test("imputa tolva primero, luego préstamo (aunque el préstamo sea más antiguo)", () => {
  const plan = planificarRecuperacion({
    parteLocal: "200.00",
    porcentajeRecuperacion: 100,
    creditos: [
      prestamo("p1", "100.00", "2026-01-01"),
      tolva("t1", "50.00", "2026-05-01"),
    ],
  });
  // objetivo 200, saldo total 150 → recuperado 150; tolva primero.
  assertEquals(plan.recuperado_total, "150.00");
  assertEquals(plan.pagado_local, "50.00");
  assertEquals(plan.asignaciones, [
    { credito_id: "t1", importe: "50.00" },
    { credito_id: "p1", importe: "100.00" },
  ]);
});

Deno.test("FIFO entre préstamos: la deuda más antigua primero", () => {
  const plan = planificarRecuperacion({
    parteLocal: "100.00",
    porcentajeRecuperacion: 100,
    creditos: [
      prestamo("nuevo", "80.00", "2026-05-01"),
      prestamo("viejo", "40.00", "2026-01-01"),
    ],
  });
  assertEquals(plan.recuperado_total, "100.00");
  assertEquals(plan.asignaciones, [
    { credito_id: "viejo", importe: "40.00" },
    { credito_id: "nuevo", importe: "60.00" },
  ]);
});

Deno.test("orden manual antepone los créditos indicados", () => {
  const plan = planificarRecuperacion({
    parteLocal: "100.00",
    porcentajeRecuperacion: 100,
    creditos: [
      prestamo("nuevo", "80.00", "2026-05-01"),
      prestamo("viejo", "40.00", "2026-01-01"),
    ],
    orden: ["nuevo"],
  });
  // 'nuevo' primero (80), luego 'viejo' (20).
  assertEquals(plan.recuperado_total, "100.00");
  assertEquals(plan.asignaciones, [
    { credito_id: "nuevo", importe: "80.00" },
    { credito_id: "viejo", importe: "20.00" },
  ]);
});
